sotodlib/hwp/hwp.py

```diff
@@ -0,0 +1,146 @@
+    band: MF1=0, MF2=1
+    loading: intensity [Kcmb]
+    inc_ang: incident angle [deg]
+    Return ---
+    Adding aman.hwpss
+    """
+    if 'hwpss' in aman.keys(): aman.move('hwpss','')
+    aman.wrap_new( 'hwpss', ('dets', 'samps'))
+    hwp_angle = np.mod(2*np.pi * aman.timestamps * hwp_freq, 2*np.pi)
+    if not 'hwp_angle' in aman.keys(): aman.wrap('hwp_angle', hwp_angle, [(0,'samps')])
+    else: aman.hwp_angle = hwp_angle
+    amp_2f, amp_4f, phi_2f, phi_4f = I_to_P_param(inc_ang=15)
+    amp = [0, 0, amp_2f, 0, amp_4f]
+    amp_r = [1, 5, amp_2f*0.05, 1, amp_4f*0.05]
+    phi = [0, 0, phi_2f, 0, phi_4f]
+    phi_r = [0, np.deg2rad(1), phi_2f*0.1, np.deg2rad(1), phi_4f*0.1]
+    nf = len(amp)
+    for n in [2,4]:
+        amp_rnd = np.random.normal(loc=amp[n], scale=amp_r[n], size=(aman.dets.count,))
+        phase_rnd = np.mod(np.random.normal(loc=phi[n], scale=phi_r[n],size=(aman.dets.count,)), 2*np.pi)
+        aman.hwpss += amp_rnd[:,None]*np.cos(n*aman.hwp_angle + phase_rnd[:,None])
+                                                 
+def func_hwpss_2f4f(x,A2,A4,phi2,phi4):
+    return A2*np.cos(2*x+phi2) + A4*np.cos(4*x+phi4)
+    
+def func_hwpss_2f(x,A2,phi2):
+    return A2*np.cos(2*x+phi2)
+                                                 
+def func_hwpss_4f(x,A4,phi4):
+    return A4*np.cos(4*x+phi4)
+
+def func_hwpss_all(x,A1,A2,A3,A4,A5,phi1,phi2,phi3,phi4,phi5):
+    return A1*np.cos(x+phi1) + A2*np.cos(2*x+phi2) + A3*np.cos(3*x+phi3) + A4*np.cos(4*x+phi4) + A5*np.cos(5*x+phi5)
+
+def ext_hwpss(aman, hwp_angle=None, mode=0, name = 'ext_hwpss', signal='signal'):    
+    """ 
+    ** HWPSS extraction function **
+    Args ---
+    aman: AxisManager
+    mode: 0 = 2f&4f, 1 = 2f only, 2 = 4f only, 3 = all up to 5f
+    name: aman output name
+    signal: target timestream data
+    Return ---
+    Adding aman.hwpss
+    """
+    filt = tod_ops.filters.high_pass_sine2(cutoff=0.1)
+    signal_prefilt = np.array(tod_ops.fourier_filter(aman, filt, detrend=None, signal_name=signal))
+    
+    bins=361
+    hwpss_denom = np.histogram(aman.hwp_angle, bins=bins, range=[0, 2*np.pi])[0]
+    hwp_angle_bins = np.linspace(0,2*np.pi,bins)
+    hwpss=[]
+    for i in range(aman.dets.count):
+        hwpss.append(np.histogram(aman.hwp_angle, bins=bins, range=[0, 2*np.pi], weights=signal_prefilt[i])[0] / np.where(hwpss_denom==0, 1, hwpss_denom))
+        idx = np.argwhere(hwpss[i]!=0).flatten()
+        if i == 0: hwp_angle_bins = hwp_angle_bins[idx]
+        hwpss[i] = hwpss[i][idx]
+    
+    if 'ext_hwpss' in aman.keys(): aman.move('ext_hwpss','')
+    aman.wrap_new( 'ext_hwpss', ('dets', 'samps'))
+    if mode == 0: func = func_hwpss_2f4f
+    elif mode == 1: func = func_hwpss_2f
+    elif mode == 2: func = func_hwpss_4f
+    elif mode == 3: func = func_hwpss_all
+    for i in range(aman.dets.count):
+        popt, pcov = curve_fit(func, hwp_angle_bins, hwpss[i], maxfev=1000000)
+        aman.ext_hwpss[i] += func(aman.hwp_angle, *popt)
```

Comment:
I think you want this to be an `=` instead of a `+=`? Technically `wrap_new` should fill with zeros but this is something you're just calculating right?

Reply:
Thanks, fixed
https://github.com/simonsobs/sotodlib/blob/80b784143dd5fb3b49f965c6ae90ed83a1962b91/sotodlib/hwp/hwp.py#L52